File: Classes/Progress Bar/TSSTInfoWindow.m
//
//  TSSTInfoWindow.m
//  SimpleComic
//
//  Created by Alexander Rauchfuss on 7/15/07.
//  Copyright 2007 Dancing Tortoise Software. All rights reserved.
//

#import "TSSTInfoWindow.h"
#import "TSSTImageUtilities.h"
#import "Simple_Comic-Swift.h"

@implementation TSSTInfoWindow


- (instancetype)initWithContentRect:(NSRect)contentRect styleMask:(NSUInteger)aStyle backing:(NSBackingStoreType)bufferingType defer:(BOOL)flag
{
    self = [super initWithContentRect: contentRect styleMask: NSBorderlessWindowMask backing: bufferingType defer: flag];
    if(self)
    {
        [self setOpaque: NO];
        [self setIgnoresMouseEvents: YES];
    }
    return self;
}


- (void)caretAtPoint:(NSPoint)point size:(NSSize)size withLimitLeft:(CGFloat)left right:(CGFloat)right
{
    CGFloat limitWidth = right - left;
    CGFloat relativePosition = (point.x - left) / limitWidth;
    CGFloat offset = size.width * relativePosition;
	NSRect frameRect = NSMakeRect( point.x - offset - 10, point.y, size.width + 20, size.height + 25);
	
	[[self contentView] setCaretPosition: offset + 10];
    [self setFrame: frameRect display: YES animate: NO];
	[self invalidateShadow];
}


- (void)centerAtPoint:(NSPoint)center
{
    NSRect frame = [self frame];
    [self setFrameOrigin: NSMakePoint(center.x - NSWidth(frame) / 2, center.y - NSHeight(frame) / 2)];
	[self invalidateShadow];
}


- (void)resizeToDiameter:(CGFloat)diameter
{
    NSRect frame = [self frame];
	NSPoint center = NSMakePoint(NSMinX( frame ) + NSWidth(frame) / 2, NSMinY( frame ) + NSHeight(frame) / 2);
	
	[self setFrame: NSMakeRect(center.x - diameter / 2,  center.y - diameter / 2, diameter, diameter) 
		   display: YES 
		   animate: NO];
}


@end



@implementation TSSTCircularImageView



- (void)drawRect:(NSRect)rect
{
	NSRect bounds = [self bounds];
	[[NSColor clearColor] set];
    NSRectFill(bounds);
    if([self image])
    {
		NSBezierPath * circle = [NSBezierPath bezierPathWithOvalInRect: NSInsetRect(bounds,5,5)];
        [[NSColor whiteColor] set];
        [circle fill];
        [circle addClip];
		[super drawRect: rect];
    }
}



@end

File: Classes/Progress Bar/TSSTInfoWindow.swift
//
//  TSSTInfoWindow.swift
//  SimpleComic
//
//  Created by C.W. Betts on 1/17/16.
//  Copyright © 2016 Dancing Tortoise Software. All rights reserved.
//

import Cocoa

/// This panel subclass is used by both the loupe,
/// and the speech bubble styled page preview.
class TSSTInfoWindow: NSPanel {
	override init(contentRect: NSRect, styleMask aStyle: NSWindow.StyleMask, backing bufferingType: NSWindow.BackingStoreType, defer flag: Bool) {
		super.init(contentRect: contentRect, styleMask: .borderless, backing: bufferingType, defer: flag)
		isOpaque = false
		ignoresMouseEvents = true
		backgroundColor = NSColor.clear
	}
	
	@objc(caretAtPoint:size:withLimitLeft:right:)
	func caret(at point: NSPoint, size: NSSize, limitLeft left: CGFloat, limitRight right: CGFloat) {
		let limitWidth = right - left
		let relativePosition = (point.x - left) / limitWidth
		let offset = size.width * relativePosition
		let frameRect = NSRect(x: point.x - offset - 10, y: point.y, width: size.width + 20, height: size.height + 25)
		(contentView as? TSSTInfoView)?.caretPosition = offset + 10
		setFrame(frameRect, display: true, animate: false)
		invalidateShadow()
	}
	
	@objc(centerAtPoint:)
	func center(at center: NSPoint) {
		let frame = self.frame
		let fo = NSPoint(x: center.x - frame.width / 2, y: center.y - frame.height / 2)
		setFrameOrigin(fo)
		invalidateShadow()
	}
	
	@objc(resizeToDiameter:)
	func resize(toDiameter diameter: CGFloat) {
		let frame = self.frame
		let center = NSPoint(x: frame.minX + frame.width / 2, y: frame.minY + frame.height / 2)
		setFrame(NSRect(x: center.x - diameter / 2, y: center.y - diameter / 2, width: diameter, height: diameter),
				 display: true,
				 animate: false)
	}
}
